hail/python/hailtop/aiotools/s3asyncfs.py

```diff
@@ -147,31 +149,33 @@ def __init__(self, mpc, number: int):
         self._mpc = mpc
         self._number = number
         self._async_writable: Optional[AsyncQueueWritableStream] = None
-        self._put_task: Optional[asyncio.Task] = None
+        self._put_thread: Optional[threading.Thread] = None
 
     async def __aenter__(self) -> WritableStream:
         async_writable, blocking_readable = async_writable_blocking_readable_stream_pair()
         self._async_writable = async_writable
-        self._put_task = asyncio.create_task(
-            blocking_to_async(self._mpc._fs._thread_pool, _upload_part,
-                              self._mpc._fs._s3,
-                              self._mpc._bucket,
-                              self._mpc._name,
-                              self._number,
-                              blocking_readable,
-                              self._mpc._upload_id))
+
+        def put():
+            self._mpc._etags[self._number] = _upload_part(
+                self._mpc._fs._s3,
+                self._mpc._bucket,
+                self._mpc._name,
+                self._number,
+                blocking_readable,
+                self._mpc._upload_id)
+
+        self._put_thread = threading.Thread(target=put)
+        self._put_thread.start()
         return async_writable
 
     async def __aexit__(
             self, exc_type: Optional[Type[BaseException]] = None,
             exc_value: Optional[BaseException] = None,
             exc_traceback: Optional[TracebackType] = None) -> None:
         assert self._async_writable is not None
-        assert self._put_task is not None
-        try:
-            await self._async_writable.wait_closed()
-        finally:
-            self._mpc._etags[self._number] = await self._put_task
+        assert self._put_thread is not None
+        await self._async_writable.wait_closed()
```

Comment:
How come this is no longer `try`'d?

Reply:
So I didn't think it was necessary since the assignment to `self._value` happened on the thread, but, separately, I realized exceptions are not automatically propagated out of the thread at the `.join()`, so I added that.